File: data_structures/singlyLinkedList/singlyLinkedList.js
/**
 * Singly Linked List data strucuture implementation:
 *
 * list's elements(nodes) are coupled with each other in one direction => each node has the pointer to next one, but not to previous.
 * last element's pointer to next node equals to null
 * nodes can be added/deleted/accessed both from _head/tail and by its position in the list.
 */

/**
 * List Node class
 * stores the data and pointer to next element
 */
 class ListNode {
  constructor(data = null, next = null) {
    this.data = data;
    this.next = next;
  }
}

/**
 * SinglyLinkedList class
 * stores pointer to the list head and current list length
 */
 class SinglyLinkedList {
  constructor() {
    this._head = null;
    this.length = 0;
  }

  getHead() {
    return this._head;
  }

  addNodeToHead(nodeData) {
    const newNode = new ListNode(nodeData);
    if (!this._head && !this.length) {
      this._head = newNode;
    } else {
      const prevHead = this._head;
      this._head = newNode;
      this._head.next = prevHead;
    }
    this.length++;
  }

  removeNodeFromHead() {
    if (!this.length) return;
    const isSingleItem = this.length === 1;
    this._head = isSingleItem ? null : this._head.next;
    this.length = isSingleItem ? 0 : this.length - 1;
  }

  getLastNode() {
    if (!this.length && !this._head) return null;

    let currentNode = this._head;
    while (currentNode.next) {
      currentNode = currentNode.next;
    }
    return currentNode;
  }

  removeLastNode() {
    if (!this.length && !this._head) return null;

    let currentNode = this._head;
    let prevNode = null;
    while (currentNode.next) {
      prevNode = currentNode;
      currentNode = currentNode.next;

      if (!currentNode.next) {
        prevNode.next = null;
        this.length--;
        return currentNode;
      }
    }
  }

  getNodeByPosition(position) {
    if (this.length < position || !this.length) return null;
    if (position === 0) return this._head;
    if (position === this.length - 1) return this.getLastNode();
    return this._loopListByPosition(
      position,
      (currentNode, currentPosition) => {
        const isCountMatch = currentPosition === position;
        return {
          done: isCountMatch,
          node: currentNode,
        };
      }
    );
  }

  addNodeToPosition(nodeData, position) {
    if (position > this.length) return null;
    if (position === 0) {
      this.addNodeToHead(nodeData);
      return this._head;
    }

    return this._loopListByPosition(
      position,
      (currentNode, currentPosition) => {
        if (currentPosition + 1 === position) {
          const newNode = new ListNode(nodeData);
          newNode.next = currentNode.next;
          currentNode.next = newNode;
          this.length++;
          return { done: true, node: newNode };
        }
        return { done: false, node: currentNode };
      }
    );
  }

  removeNodeByPosition(position) {
    if (position > this.length || !this.length) return null;
    if (position === 0) {
      const removedNode = this._head;
      this.removeNodeFromHead();
      return removedNode;
    }

    return this._loopListByPosition(
      position,
      (currentNode, currentPosition) => {
        if (currentPosition + 1 === position) {
          const _removedNode = currentNode.next;
          currentNode.next = currentNode.next ? currentNode.next.next : null;
          this.length--;
          return { done: true, node: _removedNode };
        }
        return { done: false, node: currentNode };
      }
    );
  }

  clearList() {
    this._head = null
    this.length = 0
  }

  // private methods
  _loopListByPosition(position, actionCallback) {
    let currentNode = this._head;
    let currentPosition = 0;
    while (currentPosition <= position) {
      const { done, node } = actionCallback(currentNode, currentPosition);
      if (done) return node;
      currentNode = node.next;
      currentPosition++;
    }
  }
}


module.exports = SinglyLinkedList

File: data_structures/singlyLinkedList/singly_linked_list.py
'''
Singly Linked List data strucuture implementation:

list's elements(nodes) are coupled with each other in one direction => each node has the pointer to next one, but not to previous.
last element's pointer to next node equals to null
nodes can be added/deleted/accessed both from _head/tail and by its position in the list.
'''

from operator import itemgetter


class ListNode:
    '''
    List Node class
    stores the data and pointer to next element
    '''

    def __init__(self, data=None, next=None):
        self.data = data
        self.next = next


class SinglyLinkedList:
    '''
    SinglyLinkedList class
    stores pointer to the list head and current list length
    '''

    def __init__(self):
        self.__head = None
        self.list_length = 0

    @property
    def head(self):
        return self.__head

    def add_node_to_head(self, node_data):
        new_node = ListNode(node_data)
        if not self.__head and not self.list_length:
            self.__head = new_node
        else:
            prev_head = self.__head
            self.__head = new_node
            self.__head.next = prev_head
        self.list_length += 1

    def remove_node_from_head(self):
        if not self.list_length:
            return
        is_single_item = self.list_length == 1
        self.__head = None if is_single_item else self.__head.next
        self.list_length = 0 if is_single_item else self.list_length - 1

    def get_last_node(self):
        if not self.list_length and not self.__head:
            return None
        current_node = self.__head
        while current_node.next:
            current_node = current_node.next
        return current_node

    def remove_last_node(self):
        if not self.list_length and not self.__head:
            return None
        current_node = self.__head
        prev_node = None
        while current_node.next:
            prev_node = current_node
            current_node = current_node.next
            if not current_node.next:
                prev_node.next = None
                self.list_length -= 1
                return current_node

    def get_node_by_position(self, position):
        if self.list_length < position or not self.list_length:
            return None
        if position == 0:
            return self.__head
        if position == self.list_length - 1:
            return self.get_last_node()

        def _action_callback(current_node, current_position):
            is_count_match = current_position == position
            return {"done": is_count_match, "node": current_node}
        return self.__loop_list_by_position(position, _action_callback)

    def add_node_to_position(self, node_data, position):
        if position > self.list_length:
            return None
        if position == 0:
            self.add_node_to_head(node_data)
            return self.__head

        def _action_callback(current_node, current_position):
            if current_position + 1 == position:
                new_node = ListNode(node_data)
                new_node.next = current_node.next
                current_node.next = new_node
                self.list_length += 1
                return {"done": True, "node": new_node}
            return {"done": False, "node": current_node}
        return self.__loop_list_by_position(position, _action_callback)

    def remove_node_by_position(self, position):
        if position > self.list_length or not self.list_length:
            return None
        if position == 0:
            removed_node = self.__head
            self.remove_node_from_head()
            return removed_node

        def _action_callback(current_node, current_position):
            if current_position + 1 == position:
                _removed_node = current_node.next
                current_node.next = current_node.next.next if current_node.next else None
                self.list_length -= 1
                return {"done": True, "node": _removed_node}
            return {"done": False, "node": current_node}
        return self.__loop_list_by_position(position, _action_callback)

    def clear_list(self):
        self.__node = None
        self.list_length = 0

    def __loop_list_by_position(self, position, action_callback):
        current_node = self.__head
        current_position = 0

        while current_position <= position:
            done, node = itemgetter("done", "node")(
                action_callback(current_node, current_position))
            if done:
                return node
            current_node = node.next
            current_position += 1
